kernel/proc.c

```diff
@@ -445,21 +525,31 @@ void
 scheduler(void)
 {
   struct proc *p;
+  struct proc_list* cur_proc;
   struct cpu *c = mycpu();
   
   c->proc = 0;
   for(;;){
     // Avoid deadlock by ensuring that devices can interrupt.
     intr_on();
+    // If we are on the first cpu clear the list of procs
+    if (unused_count > NPROC_KSTACK / 2) {
```

Comment:
Look quite readable to me and I strongly believe that this value should depend on NPROC_KSTACK / 2, since it is the only point of memory leaks in my system